src/dynamo0p3.py

```diff
@@ -1209,33 +1266,174 @@ def initialise_dofmaps(self, parent):
+        if self._unique_cbanded_maps:
+            parent.add(CommentGen(parent, ""))
+            parent.add(CommentGen(parent,
+                                  " Look-up required column-banded dofmaps"))
+            parent.add(CommentGen(parent, ""))
+
+            for dmap, cma in self._unique_cbanded_maps.items():
+                parent.add(AssignGen(parent, pointer=True, lhs=dmap,
+                                     rhs=cma[0].proxy_name_indexed +
+                                     "%column_banded_dofmap_"+cma[1]))
+
+        if self._unique_indirection_maps:
+            parent.add(CommentGen(parent, ""))
+            parent.add(CommentGen(parent,
+                                  " Look-up required CMA indirection dofmaps"))
+            parent.add(CommentGen(parent, ""))
 
-        for dmap, field in self._unique_fs_maps.items():
-            parent.add(AssignGen(parent, pointer=True, lhs=dmap,
-                                 rhs=field.proxy_name_indexed +
-                                 "%" + field.ref_name() +
-                                 "%get_whole_dofmap()"))
+            for dmap, cma in self._unique_indirection_maps.items():
+                parent.add(AssignGen(parent, pointer=True, lhs=dmap,
+                                     rhs=cma[0].proxy_name_indexed +
+                                     "%indirection_dofmap_"+cma[1]))
 
     def declare_dofmaps(self, parent):
         ''' Declare all unique function space dofmaps as pointers to
         integer arrays of rank 2. The declarations are added as
         children of the supplied parent argument. This must be an
         appropriate f2pygen object. '''
         from f2pygen import DeclGen
+
+        # Function space dofmaps
         decl_map_names = \
             [dmap+"(:,:) => null()" for dmap in self._unique_fs_maps]
 
         if decl_map_names:
             parent.add(DeclGen(parent, datatype="integer", pointer=True,
                                entity_decls=decl_map_names))
 
+        # Column-banded dofmaps
+        decl_bmap_names = \
+            [dmap+"(:,:) => null()" for dmap in self._unique_cbanded_maps]
+        if decl_bmap_names:
+            parent.add(DeclGen(parent, datatype="integer", pointer=True,
+                               entity_decls=decl_bmap_names))
+
+        # CMA operator indirection dofmaps
+        decl_ind_map_names = \
+            [dmap+"(:) => null()" for dmap in self._unique_indirection_maps]
+        if decl_ind_map_names:
+            parent.add(DeclGen(parent, datatype="integer", pointer=True,
+                               entity_decls=decl_ind_map_names))
+
+
+class DynInvokeCMAOperators(object):
+    ''' Holds all information on the CMA operators required by an invoke '''
+
+    cma_same_fs_params = ["nrow", "bandwidth", "alpha",
+                          "beta", "gamma_m", "gamma_p"]
+    cma_diff_fs_params = ["nrow", "ncol", "bandwidth", "alpha",
```

Comment:
This is the order that the parameters are passed in to the kernels so it matters. I've added a couple of comments.